### PR TITLE
intellij-plugin-verifier integration

### PR DESCRIPTION
**Pull request by hsz**
Implementation for #385

### CODE REVIEW

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+        project.javaexec {
+            classpath = project.files(verifierPath)
+            main = "com.jetbrains.pluginverifier.PluginVerifierMain"
+            args = verifierArgs
+        }
+    }
+
+    private String getVerifierPath() {
+        project.repositories.maven { it.url = IntelliJPlugin.DEFAULT_INTELLIJ_PLUGIN_SERVICE }
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:180]: probably it makes sense to remove repo after downloading

Comment (hsz) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy]: Good point, done.

**Review by serejke — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -155,6 +156,29 @@ The following attributes are a part of the Patching DSL <kbd>patchPluginXml { ..
 | <kbd>pluginXmlFiles</kbd> is a collection of xml files to patch.                      | All `plugin.xml` files with `<idea-plugin>` |
 | <kbd>destinationDir</kbd> is a directory to store patched xml files.                  | <kbd>'${project.buildDir}/patchedPluginXmlFiles'</kbd> |
 
+### Plugin Verifier DSL
+[IntelliJ Plugin Verifier](https://github.com/JetBrains/intellij-plugin-verifier) integration task allows to check the binary compatibility of the built plugin against the specified IntelliJ IDE builds.
```
Comment (YannCebron) [inline README.md:160]: add link to https://jetbrains.org/intellij/sdk/docs/reference_guide/api_changes_list.html

File: README.md [hunk 3]
```diff
@@ -155,6 +156,29 @@ The following attributes are a part of the Patching DSL <kbd>patchPluginXml { ..
...
+
+> **TIP** For more details, examples or issues reporting, go to the [IntelliJ Plugin Verifier](https://github.com/JetBrains/intellij-plugin-verifier) repository.
+
+| **Attributes**                                                                                                                                                | **Default Value**                                      |
+| ------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------ |
+| <kbd>ides</kbd> - Array of IDEs to check, in `intellij.version` format, like: `IC-2020.2`. If not set, uses `intellij`.                                       | <kbd>["${intellij.type}-${intellij.version}"]</kbd>    |
+| <kbd>verifierVersion</kbd> - IntelliJ Plugin Verifier version, by default uses the latest available.                                                          | <kbd>latest</kbd>                                      |
```
Comment (YannCebron) [inline README.md:169]: add recommendation not to override this unless You Know What You're Doing (tm)

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    void setSubsystemsToCheck(String subsystemsToCheck) {
+        this.subsystemsToCheck = subsystemsToCheck
+    }
+
+    @TaskAction
+    void runPluginVerifier() {
+        def extension = project.extensions.findByType(IntelliJPluginExtension)
+        def pluginFileName = "${extension.pluginName}-${project.version}"
```
[the remark this answers is not in the document]
Comment (hsz) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:216]: @zolotov Can you assist with that? WDYT about the current approach and Yann's concerns?

[the remark this answers is not in the document]
Comment (zolotov) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:216]: agree with Yann. You should accept plugin file outside. It will fix the problem with the name + it makes your task more incremental (no check for unmodified plugin) + it will be able to run verification on several plugins. See PublishTask for reference

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+
+        FailureLevel(String testValue) {
+            this.testValue = testValue
+        }
+    }
+
+    private EnumSet<FailureLevel> failureLevel = EnumSet.of(FailureLevel.INVALID_PLUGIN)
+    private List<String> ides = new ArrayList<String>()
+    private String verifierVersion = VERIFIER_VERSION_LATEST
+    private String verificationReportsDir = "${project.buildDir}/reports/pluginsVerifier"
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:47]: the default values should be set by convention, not by field value

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+        this.failureLevel = EnumSet.of(failureLevel)
+    }
+
+    @Input
+    @Optional
+    List<String> getIdes() {
+        return ides
+    }
+
+    void setIdes(List<String> ides) {
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:79]: `ides` is a bit cryptic. ideVersions is better + some JavaDoc won't hurt

File: src/docs/configuration.md.templar [hunk 7]
```diff
@@ -299,6 +299,31 @@ The following attributes are a part of the Patching DSL <kbd>patchPluginXml { ..
...
+| <kbd>verificationReportsDir</kbd> - The path to directory where verification reports will be saved.                                                            | <kbd>${project.buildDir}/reports/pluginsVerifier</kbd> |
+| <kbd>jbrVersion</kbd> - JBR version used by the Verifier.                                                                                                      | none                                                   |
+| <kbd>runtimeDir</kbd> - The path to directory containing Java runtime jar, overrides JBR.                                                                      | none                                                   |
+| <kbd>externalPrefixes</kbd> - The prefixes of classes from the external libraries.                                                                             | none                                                   |
+| <kbd>pluginsToCheckAllBuilds</kbd> - The plugin IDs to check ALL compatible plugin builds.                                                                     | none                                                   |
```
Comment (zolotov) [inline src/docs/configuration.md.templar:318]: I don't understand that. What should I set here? Considering it has a String type

File: src/docs/configuration.md.templar [hunk 8]
```diff
@@ -299,6 +299,31 @@ The following attributes are a part of the Patching DSL <kbd>patchPluginXml { ..
...
+| <kbd>failureLevel</kbd> - Defines the verification level at which task should fail. Can be set as `FailureLevel` enum or `EnumSet<FailureLevel>`.              | <kbd>FailureLevel.INVALID_PLUGIN</kbd>                 |
+| <kbd>verificationReportsDir</kbd> - The path to directory where verification reports will be saved.                                                            | <kbd>${project.buildDir}/reports/pluginsVerifier</kbd> |
+| <kbd>jbrVersion</kbd> - JBR version used by the Verifier.                                                                                                      | none                                                   |
+| <kbd>runtimeDir</kbd> - The path to directory containing Java runtime jar, overrides JBR.                                                                      | none                                                   |
+| <kbd>externalPrefixes</kbd> - The prefixes of classes from the external libraries.                                                                             | none                                                   |
```
Comment (zolotov) [inline src/docs/configuration.md.templar:317]: smae as pluginsToCheckAllBuilds

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    void setSubsystemsToCheck(String subsystemsToCheck) {
+        this.subsystemsToCheck = subsystemsToCheck
+    }
+
+    @TaskAction
+    void runPluginVerifier() {
+        def extension = project.extensions.findByType(IntelliJPluginExtension)
+        def pluginFileName = "${extension.pluginName}-${project.version}"
+
+        if (!project.file("${project.buildDir}/distributions/${pluginFileName}.zip").exists()) {
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:218]: I don't think you should fail, I believe you should `SKIP`. For that you need accept plugin file and set it `SkipIfEmpty`

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+        new ByteArrayOutputStream().withStream { os ->
+            project.javaexec {
+                classpath = project.files(verifierPath)
+                main = "com.jetbrains.pluginverifier.PluginVerifierMain"
+                args = verifierArgs
+                standardOutput = os
+            }
+
+            def output = os.toString()
+            println output
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:247]: ?

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+            for (FailureLevel level : FailureLevel.values()) {
+                if (failureLevel.contains(level) && output.contains(level.testValue)) {
+                    throw new GradleException(level.toString())
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:250]: Do we really need to fail the entire build here?

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+    private String resolveVerifierVersion() {
+        if (verifierVersion != VERIFIER_VERSION_LATEST) {
+            return verifierVersion
+        }
+
+        def url = new URL(BINTRAY_API_VERIFIER_VERSION_LATEST)
+        return new JsonSlurper().parse(url)["name"]
+    }
+
+    private String resolveRuntimeDir() {
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:278]: looks too similar to org.jetbrains.intellij.IntelliJPlugin#prepareConventionMappingsForRunIdeTask

**Review by zolotov — CHANGES_REQUESTED** [review]
please fix the comments, see `PublishTask` and `org.jetbrains.intellij.tasks.IntelliJInstrumentCodeTask` as reference

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy (hunk 11 above)
Comment (hsz) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:264]: It's supposed to fail only if the user wants it failing by setting the _failureLevel_ property. I find it useful in the case of CI to fail the job.

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy (hunk 12 above)
Comment (hsz) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:292]: Indeed, but I'm resolving different paths here - not to the binary but JRE home dir. For `builtinJbrVersion` case it's more _complicated_ than `builtinJbrVersion` in the mentioned method.

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy (hunk 10 above)
Comment (hsz) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:261]: `javaexec` above is capturing the output with OutputStream. With print, I just want to pass that again.

File: src/docs/configuration.md.templar (hunk 7 above)
Comment (hsz) [reply src/docs/configuration.md.templar]: I've copy&pasted description from the Verifier docs - it's not so clear.
Anyway, it's used for building the verification matrix: multiple plugins x IDEs; since we're going to test only one current plugin against n IDEs, it's pointless to expose that property.
Removed.

File: src/docs/configuration.md.templar (hunk 8 above)
Comment (hsz) [reply src/docs/configuration.md.templar]: As above.

**Review by YannCebron — APPROVED** [review]

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 13]
```diff
@@ -107,6 +124,38 @@ class RunPluginVerifierTask extends ConventionTask {
...
+     * @return locally installed IDEs list
+     */
+    @Input
+    List<String> getLocalPaths() {
+        return Utils.stringListInput(localPaths)
+    }
+
+    /**
+     * Sets a list of the paths to locally installed IDE distributions that should be used for verification.
+     * Property does not override {@link #ideVersions}.
```
Comment (YannCebron) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:139]: maybe reword to "in addition to those specified in ideVersions"

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 14]
```diff
@@ -107,6 +124,38 @@ class RunPluginVerifierTask extends ConventionTask {
...
+     * Accepts list of {@link String} or {@link Closure}.
+     *
+     * @param localPaths list of IDE versions
```
Comment (YannCebron) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:142]: list of paths

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 15]
```diff
@@ -107,6 +124,38 @@ class RunPluginVerifierTask extends ConventionTask {
...
+     */
+    void localPaths(List<Object> localPaths) {
```
Comment (YannCebron) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:155]: dup?

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 16]
```diff
@@ -422,10 +497,70 @@ class RunPluginVerifierTask extends ConventionTask {
             def dependency = project.dependencies.create("org.jetbrains.intellij.plugins:verifier-cli:$resolvedVerifierVersion:all@jar")
             def configuration = project.configurations.detachedConfiguration(dependency)
             return configuration.singleFile.absolutePath
+        } catch (Exception e) {
+            println e
```
Comment (YannCebron) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:508]: throw? log?

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 17]
```diff
@@ -422,10 +497,70 @@ class RunPluginVerifierTask extends ConventionTask {
...
+                Utils.debug(project, "Resolved IDE '$type-$version' path: ${dir.absolutePath}")
+                return dir.absolutePath
+            } catch (IOException e) {
+                Utils.debug(project, "Cannot download IDE '$type-$version' from $buildType channel", e)
+            }
         }
-        finally {
-            project.repositories.remove(repository)
+
+        Utils.error(project, "Cannot download IDE '$type-$version'")
```
Comment (YannCebron) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:541]: can we provide assistance?

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy [hunk 18]
```diff
@@ -422,10 +497,70 @@ class RunPluginVerifierTask extends ConventionTask {
...
+        if (!ideDir.exists()) {
+            def isBuild = version.matches("^\\d{3}\\.")
+            def ideArchive = new File(getDownloadDirectory(), "${name}.tar.gz")
+            def url = new URIBuilder(IDE_DOWNLOAD_URL)
+                    .addParameter("code", type)
+                    .addParameter("platform", "linux")
+                    .addParameter("type", buildType)
+                    .addParameter(isBuild ? "build" : "version", version)
+                    .toString()
+
```
Comment (YannCebron) [inline src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:568]: log URL

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy (hunk 15 above)
Comment (hsz) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:155]: Nope, the Gradle plugin has two setters for properties - with and without `set` prefix. I simply follow that convention.

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy (hunk 16 above)
Comment (hsz) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:508]: log

File: src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy (hunk 17 above)
Comment (hsz) [reply src/main/groovy/org/jetbrains/intellij/tasks/RunPluginVerifierTask.groovy:541]: WSYT about

```
"Cannot download IDE '$type-$version'. Please verify provided version with the available versions: https://data.services.jetbrains.com/products/releases?code=$type&type=[release|rc|eap]"
```

**Comment by zolotov** [comment]
merged manually
